docs/latex/Inhalt/03_Inhalt/06_programming_principles.tex

```diff
@@ -20,6 +20,23 @@ \subsection{Open/Closed Principle}
+
+Negativbeispiele für das Einhalten des Interface Segregation Principles finden sich bei den Interfaces \code{ProfileRepository} und \code{RecipeRepository}. Diese kombinieren jeweils den lesenden sowie schreibenden Zugriff in einem einzelnen Interface. Diese Interfaces könnten aufgeteilt werden auf jeweils ein Interface für den lesenden und ein anderes Interface für den schreibenden Zugriff. Dadurch muss ein Klient, welcher nur lesenden Zugriff benötigt, keine Abhängigkeiten auf ein Interface haben, welches schreibenden Zugriff hat. Ein Vorteil dessen könnte sein, dass dadurch die Zugriffsverwaltung leichter implementiert werden könnte. In typischen Systemen haben nicht alle Nutzer Schreibrechte, welche Leserechte haben und so könnte eine einfachere nur lesende Implementierung verwendet werden, wenn das Schreiben nicht benötigt wird.
+
+\subsection{Dependency Inversion Principle}
+Das Dependency Inversion Principle besagt, dass Details jeweils von Abstraktionen abhängen sollen. Daher werden die Schnittstellen der Implementierungen definiert und dann davon abhängig implementiert. Andere Objekte können dann von der Schnittstelle abhängig sein, ohne von konkreten Implementierungsdetails abzuhängen. Dadurch kann eine Implementierung durch eine andere ausgetauscht werden, solange dieses die gleiche Schnittstelle implementiert. Damit werden die Module voneinander entkoppelt und besser Wiederverwendbar.
+
+Das Dependency Inversion Principle wird hier zum Beispiel bei dem \code{EventCallback} eingahalten. Dieses definiert die allgemeine Schnittstelle aller Reaktionen auf ein bestimmtes Event, während in den Klassen im Package \code{org.} \code{pinkcrazyunicorn.} \code{quickie.} \code{adapters.} \code{callbacks} die konkreten Implementierungen vorhanden sind. Dadurch hat die Benutzeroberfläche in der Klasse \code{CommandLineUI} keine Abhängigkeit auf die konkreten Implementierungen, sondern kriegt diese lediglich durch Aufrufe der Methode \code{registerEvent} übergeben. Ein weiteres Beispiel findet sich bei den Repositories der Entitäten. Diese definieren jeweils alle notwendigen Methoden, die benötigt werden und die Services haben anschließend Abhängigkeiten auf die Interfaces, nicht auf die konkreten Implementierungen der Persistenz. Dadurch kann die Persistenz leicht ausgetauscht werden. Ebenfalls können die Objekte dadurch für Unit Tests durch Mock-Objekte ausgetauscht werden. Dieses wird in den Tests in den Klassen \code{ProfileServiceTest}, \code{RecipeServiceTest} und \code{MatchingServiceTest} dargestellt.
+
+Das Dependency Inversion Principle wird bei den Mapper-Klassen der Adapterschicht nicht eingehalten. Hier hängen die Klassen jeweils von konkreten Instanzen der Mapper ab. Deutlich besser wäre es gewesen, Interfaces für das Mappen von Klassen zu definieren und Implementierungen für das Mapping dieser Objekte. Dadurch wären die Implementierungen stärker entkoppelt und könnten auch in Unit Tests durch Mock-Objekte ausgetauscht werden.
```

Comment:
Interfaces für das Mappen von Klassen zu definieren und Implementierungen für das Mapping dieser Objekte -> Interfaces für das Mappen von Klassen zu definieren und diese separaten in Klassen zu implementieren. Oder so ähnlich.